### PR TITLE
[Cache] Document cache encryption using SodiumMarshaller

### LINKED ISSUE #13070: [Cache] add SodiumMarshaller
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#35019
| PR author(s) | @atailouloute
| Merged in    | 5.1-dev

### PR DESCRIPTION
**Pull request by pableu**
Fix #13070.

I'm no good at doing configs in XML and PHP, but the YAML should be fine.

#SymfonyHackday

### CODE REVIEW

File: cache.rst [hunk 1]
```diff
@@ -714,3 +714,119 @@ Clear all caches everywhere:
 .. code-block:: terminal
 
     $ php bin/console cache:pool:clear cache.global_clearer
+
+Encrypting the Cache
+--------------------
+
```
Comment (wouterj) [inline cache.rst:720]: We always add a little box that indicates that this is a new feature:

```rst
.. versionadded:: 5.2

    The ``SodiumMarshaller`` class was introduced in Symfony 5.2.
```

File: cache.rst [hunk 2]
```diff
@@ -714,3 +714,119 @@ Clear all caches everywhere:
...
+To encrypt the cache using ``libsodium``, you can use the
```
Comment (wouterj) [inline cache.rst:726]: It think we should describe a little more exact what is encrypted, the cache value? the cache item? the complete storage?

File: cache.rst [hunk 3]
```diff
@@ -714,3 +714,119 @@ Clear all caches everywhere:
...
+
+    .. code-block:: yaml
+
+        # config/packages/cache.yaml
+        services:
+            Symfony\Component\Cache\Marshaller\SodiumMarshaller:
+            decorates: cache.default_marshaller
+            arguments:
+                - ['%env(base64:CACHE_DECRYPTION_KEY)%']
+                - '@Symfony\Component\Cache\Marshaller\SodiumMarshaller.inner'
```
Comment (wouterj) [inline cache.rst]: I think this needs to be indented 4 more spaces

File: cache.rst [hunk 4]
```diff
@@ -714,3 +714,119 @@ Clear all caches everywhere:
...
+                https://symfony.com/schema/dic/symfony/symfony-1.0.xsd">
+
+            <services>
+                <service id="Symfony\Component\Cache\Marshaller\SodiumMarshaller">
+                    <factory class="Symfony\Component\Cache\Adapter\RedisAdapter" method="createConnection"/>
+                    <argument>redis://localhost</argument>
+                    <argument type="collection">
+                        <argument>env(base64:CACHE_DECRYPTION_KEY)</argument>
+                    </argument>
+                    <argument type="service" id="@Symfony\Component\Cache\Marshaller\SodiumMarshaller.inner"/>
```
Comment (wouterj) [inline cache.rst]:
```suggestion
                    <argument type="service" id="Symfony\Component\Cache\Marshaller\SodiumMarshaller.inner"/>
```

File: cache.rst [hunk 5]
```diff
@@ -714,3 +714,119 @@ Clear all caches everywhere:
...
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xmlns:framework="http://symfony.com/schema/dic/symfony"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services
+                https://symfony.com/schema/dic/services/services-1.0.xsd
+                http://symfony.com/schema/dic/symfony
+                https://symfony.com/schema/dic/symfony/symfony-1.0.xsd">
+
+            <services>
+                <service id="Symfony\Component\Cache\Marshaller\SodiumMarshaller">
```
Comment (wouterj) [inline cache.rst]:
```suggestion
                <service id="Symfony\Component\Cache\Marshaller\SodiumMarshaller"
                    decorates="cache.default_marshaller">
```

File: cache.rst [hunk 6]
```diff
@@ -714,3 +714,119 @@ Clear all caches everywhere:
...
+                    <factory class="Symfony\Component\Cache\Adapter\RedisAdapter" method="createConnection"/>
+                    <argument>redis://localhost</argument>
```
Comment (wouterj) [inline cache.rst]:
```suggestion
```

These have no corresponding lines in Yaml

File: cache.rst [hunk 7]
```diff
@@ -714,3 +714,119 @@ Clear all caches everywhere:
...
+        </container>
+
+    .. code-block:: php
+
+        // config/packages/cache.php
+        use Symfony\Component\Cache\Marshaller\SodiumMarshaller;
+
+        $container->register(SodiumMarshaller::class)
+            ->addArgument(['env(base64:CACHE_DECRYPTION_KEY)'])
+            ->addArgument(service('@Symfony\Component\Cache\Marshaller\SodiumMarshaller.inner'));
```
Comment (wouterj) [inline cache.rst:792]:
```suggestion
            ->addArgument(service('Symfony\Component\Cache\Marshaller\SodiumMarshaller.inner'));
```

**Review by wouterj — COMMENTED** [review]
Thank you!

I think we can merge both sections by adding a second example, e.g.:

```yaml
services:
    Symfony\Component\Cache\Marshaller\SodiumMarshaller:
        decorates: cache.default_marshaller
        arguments:
            - ['%env(base64:CACHE_DECRYPTION_KEY)%']
            # use multiple keys in order to rotate them
            #- ['%env(base64:CACHE_DECRYPTION_KEY)%', '%env(base64:OLD_CACHE_DECRYPTION_KEY)%']
            - '@Symfony\Component\Cache\Marshaller\SodiumMarshaller.inner'

File: cache.rst (hunk 6 above)
Comment (pableu) [reply cache.rst]: Sorry yes, I stole badly from somewhere else ;-)

File: cache.rst (hunk 3 above)
Comment (pableu) [reply cache.rst]: Thanks!

File: cache.rst (hunk 2 above)
Comment (pableu) [reply cache.rst:726]: Good idea, I added a `note` about it.

File: cache.rst (hunk 1 above)
Comment (pableu) [reply cache.rst:720]: Added (for 5.1).

**Comment by pableu** [comment]
Thanks for your suggestions, I appreciate that. 

I made a few more commits, should I rebase them all into one for a cleaner merge?

File: cache.rst [hunk 8]
```diff
@@ -714,3 +714,86 @@ Clear all caches everywhere:
...
+            // use multiple keys in order to rotate them
+            // ->addArgument(['env(base64:CACHE_DECRYPTION_KEY)', 'env(base64:OLD_CACHE_DECRYPTION_KEY)'])
+            ->addArgument(service('@Symfony\Component\Cache\Marshaller\SodiumMarshaller.inner'));
+
+To rotate your encryption keys but still be able to read existing cache entries,
+add the old encryption key to the service arguments. The first key will be used
+for reading and writing, and the additional key(s) will only be used for reading.
+
+Once all cache items encrypted with the old key have expired, you can remove
+`OLD_CACHE_DECRYPTION_KEY` completely.
```
Comment (pableu) [inline cache.rst:799]: Do you think it's worthwile explaining all that, or are the comments in the configs enough?

**Comment by wouterj** [comment]
> I made a few more commits, should I rebase them all into one for a cleaner merge?

Our merge tool takes care of merging all commits into one before merging the PR. So you don't have to worry about it (but there is nothing stopping you from doing it)

File: cache.rst (hunk 8 above)
Comment (wouterj) [reply cache.rst:799]: I like it this way!

**Review by wouterj — APPROVED** [review]

**Comment by pableu** [comment]
Cool! I'll leave it to your merge tool to do it's thing. 

I think I could also do a similar PR for #13245. Interested?

File: cache.rst [hunk 9]
```diff
@@ -714,3 +714,86 @@ Clear all caches everywhere:
...
+            ->addArgument(['env(base64:CACHE_DECRYPTION_KEY)'])
+            // use multiple keys in order to rotate them
+            // ->addArgument(['env(base64:CACHE_DECRYPTION_KEY)', 'env(base64:OLD_CACHE_DECRYPTION_KEY)'])
+            ->addArgument(service('@Symfony\Component\Cache\Marshaller\SodiumMarshaller.inner'));
+
+To rotate your encryption keys but still be able to read existing cache entries,
+add the old encryption key to the service arguments. The first key will be used
+for reading and writing, and the additional key(s) will only be used for reading.
+
+Once all cache items encrypted with the old key have expired, you can remove
```
Comment (OskarStark) [inline cache.rst:798]:
```suggestion
Once all cache items encrypted with the old key have expired, you can remove the usage of
```

File: cache.rst (hunk 8 above)
Comment (OskarStark) [inline cache.rst:799]: Please double backticks 👍🏻

**Review by OskarStark — APPROVED** [review]
Thank you, great addition to the docs 😃

**Comment by wouterj** [comment]
Thank you @pableu! This is great :smiley: 

I've fixed @OskarStark comments while merging and also did some more tweaks in 54fdf7a0771a2ea5a49886bd6cda6a7595d76742 . The diff of that commit looks a bit extreme, I mostly merged some sentences together to make this a little more concise.

> I think I could also do a similar PR for #13245. Interested?

Yes, definitely!

**Pull request merged**